File: 5/main.java
// expand around center

class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        if(s.isEmpty()) {
            return "";
        }
        
        int left = 0, right = 0;
        for(int i = 0; i < n; i++) {
            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            
            if(len > right - left) {
                left = i - (len - 1) / 2;
                right = i + len / 2;
            }
        }
        
        return s.substring(left, right + 1);
    }
    
    private int expandAroundCenter(String s, int left, int right) {
        while(left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        
        return right - left - 1;
    }
}

File: 5/main.py
# expand around center

class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if not s:
            return ""
        
        left, right = 0, 0
        for i in range(n):
            len1 = self.expandAroundCenter(s, i, i)
            len2 = self.expandAroundCenter(s, i, i + 1)
            length = max(len1, len2)
            
            if length > right - left:
                left = i - (length - 1) // 2
                right = i + length // 2
                
        return s[left:right + 1]
    
    def expandAroundCenter(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
            
        return right - left - 1
        
